src/main.rs

```diff
@@ -450,110 +593,213 @@ fn get_setup(args: &[String]) -> Setup {
     }
 
+    #[cfg(not(any(
+        feature = "alsa-backend",
+        feature = "rodio-backend",
+        feature = "portaudio-backend"
+    )))]
+    let device: Option<String> = None;
+
+    #[cfg(not(any(
+        feature = "alsa-backend",
+        feature = "rodio-backend",
+        feature = "portaudio-backend"
+    )))]
+    if matches.opt_present(DEVICE) {
+        warn!(
+            "The `--{}` / `-{}` option is not supported by the included audio backend(s), and has no effect.",
+            DEVICE, DEVICE_SHORT,
+        );
+    }
+
+    #[cfg(feature = "alsa-backend")]
     let mixer_type = matches.opt_str(MIXER_TYPE);
-    let mixer = mixer::find(mixer_type.as_deref()).expect("Invalid mixer");
+    #[cfg(not(feature = "alsa-backend"))]
+    let mixer_type: Option<String> = None;
+
+    let mixer = mixer::find(mixer_type.as_deref()).unwrap_or_else(|| {
+        error!(
+            "Invalid `--{}` / `-{}`: {}",
+            MIXER_TYPE,
+            MIXER_TYPE_SHORT,
+            matches.opt_str(MIXER_TYPE).unwrap_or_default()
+        );
+        println!(
+            "Valid `--{}` / `-{}` values: alsa, softvol",
+            MIXER_TYPE, MIXER_TYPE_SHORT
+        );
+        println!("Default: softvol");
+        exit(1);
+    });
 
     let mixer_config = {
-        let mixer_device = match matches.opt_str("mixer-card") {
-            Some(card) => {
-                warn!("--mixer-card is deprecated and will be removed in a future release.");
-                warn!("Please use --alsa-mixer-device instead.");
-                card
-            }
-            None => matches.opt_str(ALSA_MIXER_DEVICE).unwrap_or_else(|| {
-                if let Some(ref device_name) = device {
-                    device_name.to_string()
-                } else {
-                    MixerConfig::default().device
-                }
-            }),
-        };
+        let mixer_default_config = MixerConfig::default();
 
-        let index = match matches.opt_str("mixer-index") {
-            Some(index) => {
-                warn!("--mixer-index is deprecated and will be removed in a future release.");
-                warn!("Please use --alsa-mixer-index instead.");
-                index
-                    .parse::<u32>()
-                    .expect("Mixer index is not a valid number")
+        #[cfg(feature = "alsa-backend")]
+        let device = matches.opt_str(ALSA_MIXER_DEVICE).unwrap_or_else(|| {
+            if let Some(ref device_name) = device {
+                device_name.to_string()
+            } else {
+                mixer_default_config.device.clone()
             }
-            None => matches
-                .opt_str(ALSA_MIXER_INDEX)
-                .map(|index| {
-                    index
-                        .parse::<u32>()
-                        .expect("Alsa mixer index is not a valid number")
+        });
+
+        #[cfg(not(feature = "alsa-backend"))]
+        let device = mixer_default_config.device;
```

Comment:
Sorry if this was in before and I missed it in my earlier review. I'm thinking you could do `let out device = mixer_default_config.device` before the `#[cfg(feature = "alsa-backend")]` pragma and then only override device when a mixer supports that. It might cut down on a couple of pragma's, and make it easier to add a new mixer in the future (you'd only the need `#[cfg(feature = "new-mixer-thingy)]`

Reply:
What new mixer devices do you see adding in the future? Hardware volume control on Mac and Windows or something?

Reply:
At this point I don't see any, although apps downstream might add their own. Still, my suggestion might cut down a few lines of code.

Reply:
> At this point I don't see any, although apps downstream might add their own. Still, my suggestion might cut down a few lines of code.

OK sure I can do that but downstream apps wouldn't and shouldn't use our main.rs. They should use their own main and use the librespot crates and libs.

Reply:
```rust
        let mut device = mixer_default_config.device;

        #[cfg(feature = "alsa-backend")]
        device = matches.opt_str(ALSA_MIXER_DEVICE).unwrap_or_else(|| {
            if let Some(ref device_name) = device {
                device_name.to_string()
            } else {
                mixer_default_config.device.clone()
            }
        });
```

Does not work.

```
error[E0658]: attributes on expressions are experimental
   --> src/main.rs:715:9
    |
715 |         #[cfg(feature = "alsa-backend")]
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: see issue #15701 <https://github.com/rust-lang/rust/issues/15701> for more information

error: removing an expression is not supported in this position
   --> src/main.rs:715:9
    |
715 |         #[cfg(feature = "alsa-backend")]
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0308]: mismatched types
   --> src/main.rs:717:20
    |
717 |             if let Some(ref device_name) = device {
    |                    ^^^^^^^^^^^^^^^^^^^^^   ------ this expression has type `std::string::String`
    |                    |
    |                    expected struct `std::string::String`, found enum `std::option::Option`
    |
    = note: expected struct `std::string::String`
                 found enum `std::option::Option<_>`

Some errors have detailed explanations: E0308, E0658.
For more information about an error, try `rustc --explain E0308`.
```

Reply:
I also think that it's more explicit the way it is.

Reply:
Did you forget a `let` though after that pragma?
If it takes too much time then don't bother.

Reply:
Derp...

Reply:
```rust
        let device = mixer_default_config.device;

        #[cfg(feature = "alsa-backend")]
        let device = matches.opt_str(ALSA_MIXER_DEVICE).unwrap_or_else(|| {
            if let Some(ref device_name) = device {
                device_name.to_string()
            } else {
                mixer_default_config.device.clone()
            }
        });
```

Doesn't work either.

```
   --> src/main.rs:717:20
    |
717 |             if let Some(ref device_name) = device {
    |                    ^^^^^^^^^^^^^^^^^^^^^   ------ this expression has type `std::string::String`
    |                    |
    |                    expected struct `std::string::String`, found enum `std::option::Option`
    |
    = note: expected struct `std::string::String`
                 found enum `std::option::Option<_>`
```

Because:

```rust
        let device = mixer_default_config.device; #<== Is a String.

        #[cfg(feature = "alsa-backend")]
        let device = matches.opt_str(ALSA_MIXER_DEVICE).unwrap_or_else(|| {
            if let Some(ref device_name) = device { #<== This is testing against the 1st "let device" not if device_name is Some.
                device_name.to_string()
            } else {
                mixer_default_config.device.clone()
            }
        });
```

Reply:
Can we agree to cross this bridge when/if it comes. I'd really like to get this merged.

Reply:
Let me know when you want me to squash this.

Reply:
Yes I agree. You can squash if you want to or we can have GitHub do a squash & merge. Let's merge this! 👍

Reply:
I'll squash it. Give me a sec.

Reply:
All squished and ready as soon as it passes CI.